### ISSUE TITLE
Unity Beta 3 dependency dlls

### ISSUE THREAD

**Issue by leith-bartrich** [couchbase/couchbase-lite-net#508]
By default, importing the Newtonsoft.Json.dll that comes with the Unity Beta 3 release causes an Unhandled Exception in Unity.  This exception in turn shows that Unity can't import the dll.  It's throwing a TypeLoadException.

![image](https://cloud.githubusercontent.com/assets/974170/9978146/ecdf1e1c-5eef-11e5-947d-79f3239e02ea.png)

In a broader sense though, it's very problematic that you'd even distribute a Newtonsoft.Json.dll in this manner.  It's well known that in Unity, typical Json.Net distributions don't work.  They have AOT issues and other platform issues.  There is an asset in the Unity Asset Store that is widely used to avoid this problem.

https://www.assetstore.unity3d.com/en/#!/content/11347

It's very problematic for you to be distributing a dll at all for a 3rd party dependency that is so widely used, and also doesn't work.  And even the other 3rd party dependencies that do import properly are potentially in conflict with other popular assets that use the same libraries in some manner.

I'd strongly recommend you find a way to distribute your code such that Unity development workflow isn't completely hosed by your dependence on 3rd party dlls and exclusive control over how they're packaged.

This is why most complex packages in Unity are distributed as source, not as dlls, where possible.

Yes, this is a dll-hell problem.  Yes, it's partially Unity's fault for how they designed their platform.  You still need to release code that can work with it as developers use it.  Or don't bother.

**Comment by leith-bartrich** [comment]
FYI the Exception can be solved by setting the Api Compatibility Level to .Net 2.0 in the player settings.  This does not appear to be documented in the readme but can be gleaned from browsing notes on older releases.  But the larger issue of 3rd party binary dependencies stands.

**Comment by borrrden** [comment]
Couchbase Lite is 100% open source (any deps are also open source projects, but in the case of Unity3D they are mostly forked projects).  The "release" is simply a convenience for people who don't want to spend time building it.  Most of the deps I have on the repo have been turned into submodules since Unity has a hard time handling regular .NET code so I am not sure what it is you want me to do.  The JSON .NET used in the project is modified to avoid the AOT issues.  I'll make sure that our next release includes the note about the API compatibility settings.

**Comment by leith-bartrich** [comment]
To start, you could document which version of JSON.NET you built and with what configuration.  Your project references a pre-built DLL that doesn't exist in the repo or submodule, but presumably comes from one of the included solutions in the submodule with a specific configuration.

**Comment by borrrden** [comment]
> Your project refernces a pre-built DLL

Try checking out the `release/unity-ga` branch.  This has many Unity specific changes and workarounds but has not been merged into master since we are blocked on [Issue 481](https://github.com/couchbase/couchbase-lite-net/issues/481) (And by extension, [Unity issue 718885](http://issuetracker.unity3d.com/issues/il2cpp-socket-i-slash-o-hanging-with-il2cpp)).  There will be better parity once this issue is resolved and master receives the unity-specific changes (We've only gotten a vague "We might be able to look at it in the first week of October" from Unity).  It has a project reference set up for our forked version of JSON.NET (which is based on the 7.0.1 release).

**Comment by zgramana** [comment]
We'll take your comments into consideration as we prepare the packaging for GA.
